unstruct/example_test.go

```diff
@@ -0,0 +1,61 @@
+package unstruct_test
+
+import (
+	"fmt"
+
+	"k8s.io/utils/unstruct"
+
+	"gopkg.in/yaml.v2"
+)
+
+func Example() {
+	u := unstruct.New(map[string]interface{}{
+		"int": 1,
+		"list": []interface{}{
+			"A string.",
+			2,
+		},
+	})
+
+	u.Map().Field("int").Set("Number one")
```

Comment:
These examples are deceptively clean because they omit all error handling.  Once you add errors you get something more like:

```
m := u.Map()
if m == nil {
  return fmt.Errorf("expected a map")
}
f := m.Field("int")
if f == nil {
  return fmt.Errorf("expected a field 'int'")
}
f.Set("Number one")
```

and so on. Is that really any better than:

```
m, ok := u.(map[string]interface{})
if !ok {
  return fmt.Errorf("expected a map")
}
if _, found := m["int"]; !found {
  return fmt.Errorf("expected a field 'int'")
}
f["int"] = "Number one"
```

One way is idiomatic and understandable by any Go programmer (and net 1 LOC shorter).  The other way is a couple hundred LOC library, brand new, that people will have to learn.

Convince me it is actually worth doing?  I am not seeing it...

Reply:
```
m := u.Map()
if m == nil {
  return fmt.Errorf("expected a map")
}
f := m.Field("int")
if f == nil {
  return fmt.Errorf("expected a field 'int'")
}
f.Set("Number one")
```

This can actually be slightly simplified to (4 loc shorter, yay ;-)):
```
m := u.Map()
if m == nil {
  return fmt.Errorf("expected a map")
}
f := m.Field("int").Set("Number one")
```
Since `m.Field("int")` creates the field for you.

The plan is to create this abstraction and tools that can be re-used with various implementations. For example, one of the end goal is to tie this "unstruct" values with the openapi definition. So that you can walk a json and its type definition at the same time. 

I'm also building a "jsonpath" engine that let's you find values and/or specific openapi types. This would probably be the recommended way to walk a json/yaml if you don't know if fields are present or of the proper type, since it simplifies error handling. Also, that engine allows you to use "selectors" as "predicates" such as `$.store.book[?(@.price < 10)]` [1]. Which means that at this point you know the "price" field exists for each "book" value that you have (this removes the need to check that "book" is a map, that has a field "price"). Having to re-cast everything from `interface{}` to maps to ... is kind of unnecessary.

[1]: http://goessner.net/articles/JsonPath/index.html#e2